server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -454,15 +454,22 @@ private void dispatchRequest(
         }
         // TODO: estimate streamed content size for circuit breaker,
         // something like http_max_chunk_size * avg_compression_ratio(for compressed content)
-        final int contentLength = request.isFullContent() ? request.contentLength() : 0;
+        final int contentLength = request.contentLength();
         try {
             if (handler.canTripCircuitBreaker()) {
                 inFlightRequestsBreaker(circuitBreakerService).addEstimateBytesAndMaybeBreak(contentLength, "<http_request>");
             } else {
                 inFlightRequestsBreaker(circuitBreakerService).addWithoutBreaking(contentLength);
             }
             // iff we could reserve bytes for the request we need to send the response also over this channel
-            responseChannel = new ResourceHandlingHttpChannel(channel, circuitBreakerService, contentLength, methodHandlers);
+            final var resourceHandlingChannel = new ResourceHandlingHttpChannel(
+                channel,
+                circuitBreakerService,
+                contentLength,
+                methodHandlers
+            );
+            channel.request().getHttpChannel().addCloseListener(ActionListener.running(resourceHandlingChannel::close));
```

Comment:
required for client initiated channel closure in the middle of stream request

Reply:
Unfortunately, I think this is a slightly flawed approach. It would be quite common in ES for HTTP clients to use long lived http connections. This approach means that ever request will be added to a list and never removed eventually causing memory issues. I think we will need some type of approach like used in `RestCancellableNodeClient`. A registry. 

We will probably also need to use that "registry" to cancel in progress stream handlers.

Reply:
Thank you, I missed that. Registry makes sense